### PR TITLE
Refactor dynamicvalidator code

### PR DESCRIPTION
**Pull request by olga-mir**
### Which issue this PR addresses:

https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/7124129/

### What this PR does / why we need it:

In preparation to implementing the monitor as described in Work Item above I need to make dynamicvalidator methods usable from another package. 

### Test plan for issue:

this package only uses unit tests. Existing unit tests pass.

### Is there any documentation that needs to be updated for this PR?

no

### CODE REVIEW

File: pkg/api/validate/openshiftcluster_validatedynamic.go [hunk 1]
```diff
@@ -64,42 +95,34 @@ func (*azureClaim) Valid() error {
...
 }
 
 // Dynamic validates an OpenShift cluster
 func (dv *openShiftClusterDynamicValidator) Dynamic(ctx context.Context) error {
 	// TODO: Dynamic() should work on an enriched oc (in update), and it
 	// currently doesn't.  One sticking point is handling subnet overlap
 	// calculations.
-
-	r, err := azure.ParseResourceID(dv.oc.ID)
+	err := dv.Setup(ctx)
```
Comment (olga-mir) [inline pkg/api/validate/openshiftcluster_validatedynamic.go:121]: monitor, or possibly the diagnostic GA, will not run all the checks from within the same function as it is now in `Dynamic`. Therefore the setup part also needs to be taken out as a separate method

File: pkg/api/validate/openshiftcluster_validatedynamic.go [hunk 2]
```diff
@@ -110,33 +133,27 @@ func (dv *openShiftClusterDynamicValidator) Dynamic(ctx context.Context) error {
...
 	}
 
-	err = dv.validateRouteTablePermissions(ctx, dv.fpAuthorizer, dv.fpPermissions, &vnet, api.CloudErrorCodeInvalidResourceProviderPermissions, "resource provider")
+	err = dv.ValidateRouteTablePermissionsRP(ctx, &vnetr)
 	if err != nil {
 		return err
 	}
 
-	err = dv.validateVnet(ctx, &vnet)
+	err = dv.ValidateVnet(ctx, &vnetr)
```
Comment (olga-mir) [inline pkg/api/validate/openshiftcluster_validatedynamic.go:156]: in this way the user of this package does not need to retrieve the vnet which will require setting up the required authorizers

**Comment by jim-minter** [comment]
Good hack :-)  I'm +1 on *somehow* reusing validatedynamic functionality and calling it from (e.g.) the operator rather than
duplicating.  A few key points for consideration:

1) fpAuthorizer cannot be instantiated outside of the RP and anything that uses it can't be called outside of the RP.  If we could avoid it, it would be nicer if we avoided having a struct where in some contexts fpAuthorizer was nil and some not.

2) it is almost certainly inconvenient to be using api.CloudError* outside of the context of the RP.  I don't think it makes sense to use that in the operator.

3) clientAttributes is pretty clunky as it stands.

The above makes me think that we should be separating more clearly into two pieces:

a) the RP dynamic validator, which knows about different authorizers and cares about api.CloudError*

b) the core functionality which validates (like the validateVnet function), which is passed in the appropriate refreshable.Authorizer.

Part (a) would be used only in the RP; part (b) would be shared between RP and operator.

4) As things stand the interplay between NewOpenShiftClusterDynamicValidator() and Setup() is also a bit clunky in this PR.  It would be nice to have a clearer/simpler external API to this.  For example, unexport Setup(), or move the codebase into
NewOpenShiftClusterDynamicValidator().  I think this will work itself out based on the comments above.

**Comment by olga-mir** [comment]
@jim-minter we have agreed to not implement any of this logic in the operator. These checks are intended to be used from monitor. Maybe in the future we can also add a diagnostic GA action.

**Comment by olga-mir** [comment]
Given that this code is not going to be used from the operator and the fact how deeply api.CloudError is interwoven into dynamic validator code I don't think the separation like you suggest is possible or needed. 
I agree that the clientAttributes is artificial (and the naming is completely off, I am trying to come up with another name), this is still feels a lot of improvement over passing in object's attributes into the method of the same object, not to mention the raw strings arguments.

**Comment by jim-minter** [comment]
> @jim-minter we have agreed to not implement any of this logic in the operator
Yes, you're right on that, sorry.

> Given that this code is not going to be used from the operator and the fact how deeply api.CloudError is interwoven into dynamic validator code I don't think the separation like you suggest is possible or needed.
I disagree - the monitor code should not have access to fpAuthorizer either.  I think it's worth at least sketching out.

**Comment by olga-mir** [comment]
we agreed to tackle the whole task from another angle. This PR is not needed at this stage, closing

**Pull request closed without merging**